--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -37,10 +_,24 @@
    private volatile boolean f_290367_ = false;
 
    public ServerCommonPacketListenerImpl(MinecraftServer p_299469_, Connection p_300872_, CommonListenerCookie p_300277_) {
+      this(p_299469_, p_300872_, p_300277_, new net.minecraft.server.level.ServerPlayer(p_299469_, p_299469_.m_129783_(), p_300277_.f_290628_(), p_300277_.f_290565_()));
+   }
+   public ServerCommonPacketListenerImpl(MinecraftServer p_299469_, Connection p_300872_, CommonListenerCookie p_300277_, net.minecraft.server.level.ServerPlayer player) { // CraftBukkit
       this.f_291389_ = p_299469_;
       this.f_291338_ = p_300872_;
       this.f_290835_ = Util.m_137550_();
       this.f_291042_ = p_300277_.f_291325_();
+      // CraftBukkit start - add fields and methods
+      this.player = player;
+      this.cserver = p_299469_.server;
+   }
+   protected final net.minecraft.server.level.ServerPlayer player;
+   protected final org.bukkit.craftbukkit.v1_20_R2.CraftServer cserver;
+   public boolean processedDisconnect;
+
+   public org.bukkit.craftbukkit.v1_20_R2.entity.CraftPlayer getCraftPlayer() {
+      return (this.player == null) ? null : (org.bukkit.craftbukkit.v1_20_R2.entity.CraftPlayer) this.player.getBukkitEntity();
+      // CraftBukkit end
    }
 
    public void m_7026_(Component p_300550_) {
@@ -52,6 +_,7 @@
    }
 
    public void m_295033_(ServerboundKeepAlivePacket p_299975_) {
+      PacketUtils.m_131359_(p_299975_, this, this.player.m_284548_()); // CraftBukkit
       if (this.f_291554_ && p_299975_.m_293668_() == this.f_291155_) {
          int i = (int)(Util.m_137550_() - this.f_290835_);
          this.f_291042_ = (this.f_291042_ * 3 + i) / 4;
@@ -65,8 +_,18 @@
    public void m_293596_(ServerboundPongPacket p_299461_) {
    }
 
+   // CraftBukkit start
+   private static final net.minecraft.resources.ResourceLocation CUSTOM_REGISTER = new net.minecraft.resources.ResourceLocation("register");
+   private static final net.minecraft.resources.ResourceLocation CUSTOM_UNREGISTER = new net.minecraft.resources.ResourceLocation("unregister");
    public void m_293234_(ServerboundCustomPayloadPacket p_300164_) {
-   }
+      net.minecraftforge.network.NetworkHooks.onCustomPayload(p_300164_, this.f_291338_);
+      //TODO: yoink from magma
+   }
+
+   public final boolean isDisconnected() {
+      return !this.player.joining && !this.f_291338_.m_129536_();
+   }
+   // CraftBukkit end
 
    public void m_293248_(ServerboundResourcePackPacket p_300656_) {
       PacketUtils.m_131363_(p_300656_, this, this.f_291389_);
@@ -74,13 +_,14 @@
          f_291096_.info("Disconnecting {} due to resource pack rejection", (Object)this.m_293343_().getName());
          this.m_294716_(Component.m_237115_("multiplayer.requiredTexturePrompt.disconnect"));
       }
+      this.cserver.getPluginManager().callEvent(new org.bukkit.event.player.PlayerResourcePackStatusEvent(getCraftPlayer(), org.bukkit.event.player.PlayerResourcePackStatusEvent.Status.values()[p_300656_.m_294355_().ordinal()])); // CraftBukkit
 
    }
 
    protected void m_295188_() {
       this.f_291389_.m_129905_().m_6180_("keepAlive");
       long i = Util.m_137550_();
-      if (i - this.f_290835_ >= 15000L) {
+      if (i - this.f_290835_ >= 25000L) { // CraftBukkit
          if (this.f_291554_) {
             this.m_294716_(f_290625_);
          } else {
@@ -108,6 +_,13 @@
    }
 
    public void m_295553_(Packet<?> p_300325_, @Nullable PacketSendListener p_301165_) {
+      // CraftBukkit start
+      if (p_300325_ == null || this.processedDisconnect) { // Spigot
+         return;
+      } else if (p_300325_ instanceof net.minecraft.network.protocol.game.ClientboundSetDefaultSpawnPositionPacket packet) {
+         this.player.compassTarget = org.bukkit.craftbukkit.v1_20_R2.util.CraftLocation.toBukkit(packet.f_133111_, this.getCraftPlayer().getWorld());
+      }
+      // CraftBukkit end
       boolean flag = !this.f_290367_ || !this.f_291389_.m_18695_();
 
       try {
@@ -122,12 +_,65 @@
       }
    }
 
+   // CraftBukkit start
    public void m_294716_(Component p_299122_) {
-      this.f_291338_.m_243124_(new ClientboundDisconnectPacket(p_299122_), PacketSendListener.m_243092_(() -> {
-         this.f_291338_.m_129507_(p_299122_);
+      disconnect(org.bukkit.craftbukkit.v1_20_R2.util.CraftChatMessage.fromComponent(p_299122_));
+   }
+   // CraftBukkit end
+
+   public void disconnect(String s) {
+      // CraftBukkit start - fire PlayerKickEvent
+      if (this.processedDisconnect) {
+         return;
+      }
+      if (!this.cserver.isPrimaryThread()) {
+         org.bukkit.craftbukkit.v1_20_R2.util.Waitable waitable = new org.bukkit.craftbukkit.v1_20_R2.util.Waitable() {
+            @Override
+            protected Object evaluate() {
+               ServerCommonPacketListenerImpl.this.disconnect(s);
+               return null;
+            }
+         };
+
+         this.f_291389_.processQueue.add(waitable);
+
+         try {
+            waitable.get();
+         } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+         } catch (java.util.concurrent.ExecutionException e) {
+            throw new RuntimeException(e);
+         }
+         return;
+      }
+
+      String leaveMessage = net.minecraft.ChatFormatting.YELLOW + this.player.m_6302_() + " left the game.";
+
+      org.bukkit.event.player.PlayerKickEvent event = new org.bukkit.event.player.PlayerKickEvent(this.player.getBukkitEntity(), s, leaveMessage);
+
+      if (this.cserver.getServer().m_130010_()) {
+         this.cserver.getPluginManager().callEvent(event);
+      }
+
+      if (event.isCancelled()) {
+         // Do not kick the player
+         return;
+      }
+      this.player.kickLeaveMessage = event.getLeaveMessage(); // CraftBukkit - SPIGOT-3034: Forward leave message to PlayerQuitEvent
+      // Send the possibly modified leave message
+      final Component ichatbasecomponent = org.bukkit.craftbukkit.v1_20_R2.util.CraftChatMessage.fromString(event.getReason(), true)[0];
+      // CraftBukkit end
+
+      this.f_291338_.m_243124_(new ClientboundDisconnectPacket(ichatbasecomponent), PacketSendListener.m_243092_(() -> {
+         this.f_291338_.m_129507_(ichatbasecomponent);
       }));
+      this.m_7026_(ichatbasecomponent); // CraftBukkit - fire quit instantly
       this.f_291338_.m_129540_();
-      this.f_291389_.m_18709_(this.f_291338_::m_129541_);
+      MinecraftServer minecraftserver = this.f_291389_;
+
+      java.util.Objects.requireNonNull(this.f_291338_);
+      // CraftBukkit - Don't wait
+      minecraftserver.m_6681_(this.f_291338_::m_129541_);
    }
 
    protected boolean m_293330_() {
@@ -147,5 +_,9 @@
 
    protected CommonListenerCookie m_295477_(ClientInformation p_297318_) {
       return new CommonListenerCookie(this.m_293343_(), this.f_291042_, p_297318_);
+   }
+
+   public Connection getConnection() {
+      return this.f_291338_;
    }
 }
