--- a/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
+++ b/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
@@ -8,12 +_,13 @@
    private static final int f_179799_ = 255;
 
    public ClientIntentionPacket(FriendlyByteBuf p_179801_) {
-      this(p_179801_.m_130242_(), p_179801_.m_130136_(255), p_179801_.readUnsignedShort(), ClientIntent.m_293987_(p_179801_.m_130242_()));
+      // Spigot - increase max hostName length
+      this(p_179801_.m_130242_(), p_179801_.m_130136_(Short.MAX_VALUE), p_179801_.readUnsignedShort(), ClientIntent.m_293987_(p_179801_.m_130242_()));
    }
 
    public void m_5779_(FriendlyByteBuf p_134737_) {
       p_134737_.m_130130_(this.f_134720_);
-      p_134737_.m_130070_(this.f_134721_);
+      p_134737_.m_130070_(this.f_134721_ + "\0" + net.minecraftforge.network.NetworkConstants.NETVERSION + "\0");
       p_134737_.writeShort(this.f_134722_);
       p_134737_.m_130130_(this.f_134723_.m_293527_());
    }
