--- a/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
+++ b/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
@@ -8,7 +_,8 @@
    private static final int f_179799_ = 255;
 
    public ClientIntentionPacket(FriendlyByteBuf p_179801_) {
-      this(p_179801_.m_130242_(), p_179801_.m_130136_(255), p_179801_.readUnsignedShort(), ClientIntent.m_293987_(p_179801_.m_130242_()));
+      // Spigot - increase max hostName length
+      this(p_179801_.m_130242_(), p_179801_.m_130136_(Short.MAX_VALUE), p_179801_.readUnsignedShort(), ClientIntent.m_293987_(p_179801_.m_130242_()));
    }
 
    public void m_5779_(FriendlyByteBuf p_134737_) {
